Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift

```diff
@@ -61,8 +61,29 @@ class PushEventHandlerProxyImpl: PushEventHandlerProxy {
             return
         }
 
+        // Do not call the function's completionHandler until *all* delegates have had their completion handler called.
+        // This gives apps push event handlers chance to process the push and not run risk of OS terminating before processing finishes.
+        let numberOfAsyncOperationsExpected = nestedDelegates.count
+        var numberOfAsyncOperationsCompleted = 0
+        // 2+ other push event handlers may exist in app. We need to decide if a push should be displayed or not, by combining all the results from all other push handlers.
+        // To do that, we start with Apple's default value of: do not display.
+        // If any of the handlers return result indicating push should be displayed, we return true.
+        // Apple docs: https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/usernotificationcenter(_:willpresent:withcompletionhandler:)
+        var shouldDisplayPush = false
+
         nestedDelegates.forEach { _, delegate in
-            delegate.shouldDisplayPushAppInForeground(push, completionHandler: completionHandler)
+            delegate.shouldDisplayPushAppInForeground(push, completionHandler: { delegateShouldDisplayPushResult in
+                if delegateShouldDisplayPushResult {
+                    shouldDisplayPush = true
+                }
+
+                numberOfAsyncOperationsCompleted += 1
+
+                // Do not call the function's completionHandler until *all* delegates have had their completion handler called.
+                if numberOfAsyncOperationsExpected == numberOfAsyncOperationsCompleted {
+                    completionHandler(shouldDisplayPush)
+                }
```

Comment:
There are race conditions in this code that can result in the completionHandler in the if conditions runs too early or never run

Reply:
"never run" - I suggest we cover that use case in [this thread](https://github.com/customerio/customerio-ios/pull/583#discussion_r1509511361). 

"runs too early" - Could you give steps on how to reproduce this?

Reply:
> "never run" - I suggest we cover that use case in https://github.com/customerio/customerio-ios/pull/583#discussion_r1509511361.

This has nothing to with the mentioned thread. This could happen if the race condition caused the counter to miss one count. Then it will never reach the `numberOfAsyncOperationsExpected`.

> "runs too early" - Could you give steps on how to reproduce this?

Hard to explain without lengthy explanation but let's try.
in a simple scenario where `numberOfAsyncOperationsExpected` is set to 2.
Now imagine two threads increase the counter; the first one to go past line 80 after the two threads increased the counter, will meet the condition and call the completion handler, in undetermined order. Then the second one, will call it again.

Basically, there is a race condition, all sort of shenanigans can go.

What's even worse with such simple value (int in this case) that it can't be torn apart (there are exceptions but not needed here); meaning you won't access it in a way that ends up accessing half-baked memory or getting seg_fault errors. So you will never see a crash out of that, but you would see random behavior.

Reply:
I agree with the race condition possibility here, this is why i suggested using `Dispatch Group` in my comment in the other [PR](https://github.com/customerio/customerio-ios/pull/583/files#r1510386994).

Reply:
I am on the same page with what you mean by race condition now. Thanks. 

Going off of Shahroz's suggestion, DispatchGroup could be useful here. I don't currently agree with implementing a timeout, although I do understand the value of it. I think the OS should enforce the timeout. It's important to me that we keep our code as close to OS behavior as possible for customers. 

With the timeout code removed, a DisptchGroup code snippet could look similar to this: 

```swift  
let dispatchGroup = DispatchGroup()

for (_, delegate) in nestedDelegates {
    dispatchGroup.enter()

    delegate.onPushAction(pushAction) {
        dispatchGroup.leave()
    }
}

dispatchGroup.notify(queue: .main) {
    completionHandler()
}
```

Unless I hear differently, I will update both PRs with this pattern. I will do that [after working on this ticket](https://linear.app/customerio/issue/MBL-154/process-push-event-before-we-call-3rd-party-callback-functions) to give the team more time to review this PR and give suggestions.

Reply:
Are we okay with the fact, we won't be opening deeplinks for a user if any of the `completionHandler()` is not called from any of the delegates.

Reply:
Are we okay with the fact, we won't be opening deeplinks for a user if any of the `completionHandler()` is not called from any of the delegates?

Reply:
additionally, i think @Ahmed-CIO mentioned they would not prefer to go with `DispatchGroup`, so I would wait for him to raise any concern here before we proceed with it.

Reply:
Thanks both for the farther elaboration here.
The point about deep linking is also really important one. Not opening deep linking if a delegate didn't call the completionHandler can also fire back at us.

I will read through the updates I missed and come back to this one again

Reply:
Alright, after quick catch, I also did quick refresher about DispatchGroup and my worry about causing deadlock isn't actually valid. 
The only thing that still valid is using dispatchGroup.enter() must be balanced with dispatchGroup.leave(). While in the above code this is the case and easy to reason about, this is prone to all sort of issues one used to do in the early days of manually calling alloc/release. However, this is not as a hard blocker as the point I initially raised and turns out to be wrong.
My recommendation: if it is not too much work to create a small actor that only contains the parts that needs to synced, then let's do that as it is likely to be much cleaner solution. Otherwise, go with DispatchGroup. If it helps, here is a quick non-fully-validated example of how actor could help
```swift
actor PushDelegateAsyncOperationsTracker {
    private var numberOfAsyncOperationsExpected: Int
    private var shouldShowNotification: Bool = false
    private let completionHandler: (Bool) -> Void
    
    init(_ numberOfAsyncOperationsExpected: Int,
         completionHandler: @escaping (Bool) -> Void) {
        self.numberOfAsyncOperationsExpected = numberOfAsyncOperationsExpected
        self.completionHandler = completionHandler
    }
    
    
    func onAsyncOperationExecuted(_ shouldShowNotification: Bool) {
        self.numberOfAsyncOperationsExpected -= 1
        self.shouldShowNotification = self.shouldShowNotification || shouldShowNotification
        
        if self.numberOfAsyncOperationsExpected == 0 {
            self.completionHandler(self.shouldShowNotification)
        }
    }
}

delegate.shouldDisplayPushAppInForeground(push)  { delegateShouldDisplayPushResult in
        Task {
            await operationsTracker.onAsyncOperationExecuted(delegateShouldDisplayPushResult)
        }
 })
    
```

Reply:
Wrote some code and shared it with team in Slack. After response, I'll update this PR.